### PR TITLE
Fix (some) clippy issues

### LINKED ISSUE #134: Fix remaining clippy issues
```
warning: defining a method called `from_str` on this type; consider implementing the `std::str::FromStr` trait or choosing a less ambiguous name
  --> src/wrapper/signature.rs:49:5
   |
49 | /     pub fn from_str(s: &str) -> Result<JavaType> {
50 | |         Ok(
51 | |             match parser(parse_type).parse(State::new(s)).map(|res| res.0) {
52 | |                 Ok(sig) => sig,
...  |
55 | |         )
56 | |     }
   | |_____^
   |
   = note: #[warn(should_implement_trait)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#should_implement_trait

warning: defining a method called `from_str` on this type; consider implementing the `std::str::FromStr` trait or choosing a less ambiguous name
  --> src/wrapper/signature.rs:82:5
   |
82 | /     pub fn from_str<S: AsRef<str>>(s: S) -> Result<TypeSignature> {
83 | |         Ok(match parser(parse_sig)
84 | |             .parse(State::new(s.as_ref()))
85 | |             .map(|res| res.0)
...  |
90 | |         })
91 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#should_implement_trait

warning: transmute from a reference to a reference
  --> src/wrapper/strings/ffi_str.rs:31:18
   |
31 |         unsafe { ::std::mem::transmute(self.internal.as_bytes_with_nul()) }
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*(self.internal.as_bytes_with_nul() as *const [u8] as *const wrapper::strings::ffi_str::JNIStr)`
   |
   = note: #[warn(transmute_ptr_to_ptr)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#transmute_ptr_to_ptr

warning: transmute from a reference to a reference
  --> src/wrapper/strings/ffi_str.rs:86:9
   |
86 |         ::std::mem::transmute(ffi::CStr::from_ptr(ptr))
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*(ffi::CStr::from_ptr(ptr) as *const std::ffi::CStr as *const wrapper::strings::ffi_str::JNIStr)`
   |
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#transmute_ptr_to_ptr
```

### PR DESCRIPTION
**Pull request by DarkEld3r**
## Overview

Partially fixes #134 issue. The remaining issue (`FromStr` implementation for `TypeSignature`) requires more changes (because custom `from_str` have different signature from `FromStr` trait) and I prefer to separate it.

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by automated [tests](https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#tests)
- [x] The [coding guidelines](https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [x] Public API has documentation
- [x] This change is not breaking **or** mentioned in the Changelog
- [x] The [continuous integration build](https://www.travis-ci.org/jni-rs/jni-rs) passes

### CODE REVIEW

File: src/wrapper/signature.rs [hunk 1]
```diff
@@ -44,13 +45,14 @@ pub enum JavaType {
...
-    pub fn from_str(s: &str) -> Result<JavaType> {
+impl FromStr for JavaType {
+    type Err = String;
+
+    fn from_str(s: &str) -> ::std::result::Result<Self, Self::Err> {
         Ok(
             match parser(parse_type).parse(State::new(s)).map(|res| res.0) {
                 Ok(sig) => sig,
-                Err(e) => return Err(format_error_message(&e, s).into()),
+                Err(e) => return Err(format_error_message(&e, s)),
```
Comment (alexander-irbis) [inline src/wrapper/signature.rs]: If I were a clippy, I would suggest to rewrite it like this
```rust
        parser(parse_type)
            .parse(State::new(s))
            .map(|res| res.0)
            .map_err(|e| format_error_message(&e, s))
```

File: src/wrapper/strings/ffi_str.rs [hunk 2]
```diff
@@ -28,7 +29,7 @@ impl ::std::ops::Deref for JNIString {
     type Target = JNIStr;
 
     fn deref(&self) -> &Self::Target {
-        unsafe { ::std::mem::transmute(self.internal.as_bytes_with_nul()) }
+        unsafe { &*(self.internal.as_bytes_with_nul() as *const [u8] as *const ffi_str::JNIStr) }
```
Comment (dmitry-timofeev) [inline src/wrapper/strings/ffi_str.rs]: This double "as" confuses me a little :smile:

File: src/wrapper/signature.rs [hunk 3]
```diff
@@ -44,13 +45,14 @@ pub enum JavaType {
     Method(Box<TypeSignature>),
 }
 
-impl JavaType {
-    /// Parse a type string into a JavaType enum.
-    pub fn from_str(s: &str) -> Result<JavaType> {
+impl FromStr for JavaType {
```
Comment (dmitry-timofeev) [inline src/wrapper/signature.rs]: Just a question: is it a breaking change — the fact that we implement a method from a trait/not declare it in this type?

File: src/wrapper/strings/ffi_str.rs (hunk 2 above)
Comment (alexander-irbis) [reply src/wrapper/strings/ffi_str.rs]: The first `as` converts the shared reference into the const pointer, the second `as` converts the type of the pointer. Both at the same time is too implicit and may lead to an error, I believe.

File: src/wrapper/signature.rs (hunk 3 above)
Comment (alexander-irbis) [reply src/wrapper/signature.rs]: To use the method `from_str`, there should be added a `use std::str::FromStr;` statement, in a module where `from_str` is called.

Comment (dmitry-timofeev) [reply src/wrapper/signature.rs]: Therefore it is breaking if `JavaType#from_str` is public, isn't it?

Comment (alexander-irbis) [reply src/wrapper/signature.rs]: Yes, exactly.

Comment (DarkEld3r) [reply src/wrapper/signature.rs]: I'm going to update the changelog. :sweat_smile:

**Pull request merged**